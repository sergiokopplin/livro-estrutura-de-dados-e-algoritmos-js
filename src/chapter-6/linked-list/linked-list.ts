function defaultEquals(a: any, b: any) {
  return a === b;
}

class Node {
  element: any;
  next: any;

  constructor(element: any) {
    this.element = element;
    this.next = undefined;
  }
}

export class LinkedList {
  count: number;
  head: any;
  equalsFn: Function;

  constructor(equalsFn = defaultEquals) {
    this.count = 0;
    this.head = undefined;
    this.equalsFn = equalsFn;
  }

  push(element: any): void {
    const node = new Node(element);
    let current;

    if (this.head == null) {
      this.head = node;
    } else {
      current = this.head;

      while (current.next != null) {
        current = current.next;
      }

      current.next = node;
    }

    this.count++;
  }

  removeAt(index: number): any {
    if (index >= 0 && index < this.count) {
      let current = this.head;

      if (index === 0) {
        this.head = current.next;
      } else {
        const previous = this.getElementAt(index - 1);
        current = previous.next;
        previous.next = current.next;
      }

      this.count--;
      return current.element;
    }

    return undefined;
  }

  getElementAt(index: number): any {
    if (index >= 0 && index <= this.count) {
      let node = this.head;
      for (let i = 0; i < index && node != null; i++) {
        node = node.next;
      }

      return node;
    }

    return undefined;
  }

  insert(element: any, index: number): boolean {
    if (index >= 0 && index <= this.count) {
      const node = new Node(element);

      if (index === 0) {
        const current = this.head;
        node.next = current;
        this.head = node;
      } else {
        const previous = this.getElementAt(index - 1);
        const current = previous.next;
        node.next = current;
        previous.next = node;
      }

      this.count++;
      return true;
    }

    return false;
  }
}
