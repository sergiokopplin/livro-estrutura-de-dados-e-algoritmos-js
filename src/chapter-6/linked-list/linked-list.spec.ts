import { LinkedList } from './linked-list';

describe('LinkedList', () => {
  test('Should construct correctly', () => {
    const linkedList = new LinkedList();

    expect(linkedList).toMatchSnapshot();
  });

  test('Should push', () => {
    const linkedList = new LinkedList();

    linkedList.push(1);
    linkedList.push(2);
    linkedList.push(3);

    expect(linkedList).toMatchSnapshot();
  });

  test('Should removeAt', () => {
    const linkedList = new LinkedList();

    linkedList.push(1);
    linkedList.push(2);
    linkedList.push(3);
    linkedList.push(4);
    linkedList.push(5);

    expect(linkedList).toMatchSnapshot();

    expect(linkedList.removeAt(0)).toEqual(1);
    expect(linkedList.removeAt(3)).toEqual(5);
    expect(linkedList.removeAt(10)).toEqual(undefined);

    expect(linkedList).toMatchSnapshot();
  });

  test('Should getElementAt', () => {
    const linkedList = new LinkedList();

    linkedList.push(1);
    linkedList.push(2);
    linkedList.push(3);
    linkedList.push(4);
    linkedList.push(5);

    expect(linkedList.getElementAt(0)).toMatchSnapshot();
    expect(linkedList.getElementAt(3)).toMatchSnapshot();
    expect(linkedList.getElementAt(10)).toMatchSnapshot();
  });

  test('Should insert', () => {
    const linkedList = new LinkedList();

    linkedList.push(1);
    linkedList.push(2);
    linkedList.push(3);
    linkedList.push(4);
    linkedList.push(5);

    expect(linkedList.insert(10, 3)).toBe(true);
    expect(linkedList.insert(100, 1)).toBe(true);
    expect(linkedList.insert(1000, 0)).toBe(true);
    expect(linkedList.insert(10000, 100)).toBe(false);

    expect(linkedList.getElementAt(10)).toMatchSnapshot();
  });
});
