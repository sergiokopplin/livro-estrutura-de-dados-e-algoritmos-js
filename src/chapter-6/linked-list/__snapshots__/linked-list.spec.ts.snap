// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedList Should construct correctly 1`] = `
LinkedList {
  "count": 0,
  "equalsFn": [Function],
  "head": undefined,
}
`;

exports[`LinkedList Should getElementAt 1`] = `
Node {
  "element": 1,
  "next": Node {
    "element": 2,
    "next": Node {
      "element": 3,
      "next": Node {
        "element": 4,
        "next": Node {
          "element": 5,
          "next": undefined,
        },
      },
    },
  },
}
`;

exports[`LinkedList Should getElementAt 2`] = `
Node {
  "element": 4,
  "next": Node {
    "element": 5,
    "next": undefined,
  },
}
`;

exports[`LinkedList Should getElementAt 3`] = `undefined`;

exports[`LinkedList Should getHead 1`] = `
Node {
  "element": 1,
  "next": Node {
    "element": 2,
    "next": Node {
      "element": 3,
      "next": Node {
        "element": 4,
        "next": Node {
          "element": 5,
          "next": undefined,
        },
      },
    },
  },
}
`;

exports[`LinkedList Should getHead 2`] = `
Node {
  "element": 1000,
  "next": Node {
    "element": 1,
    "next": Node {
      "element": 2,
      "next": Node {
        "element": 3,
        "next": Node {
          "element": 4,
          "next": Node {
            "element": 5,
            "next": undefined,
          },
        },
      },
    },
  },
}
`;

exports[`LinkedList Should insert 1`] = `undefined`;

exports[`LinkedList Should push 1`] = `
LinkedList {
  "count": 3,
  "equalsFn": [Function],
  "head": Node {
    "element": 1,
    "next": Node {
      "element": 2,
      "next": Node {
        "element": 3,
        "next": undefined,
      },
    },
  },
}
`;

exports[`LinkedList Should removeAt 1`] = `
LinkedList {
  "count": 5,
  "equalsFn": [Function],
  "head": Node {
    "element": 1,
    "next": Node {
      "element": 2,
      "next": Node {
        "element": 3,
        "next": Node {
          "element": 4,
          "next": Node {
            "element": 5,
            "next": undefined,
          },
        },
      },
    },
  },
}
`;

exports[`LinkedList Should removeAt 2`] = `
LinkedList {
  "count": 3,
  "equalsFn": [Function],
  "head": Node {
    "element": 2,
    "next": Node {
      "element": 3,
      "next": Node {
        "element": 4,
        "next": undefined,
      },
    },
  },
}
`;
